### PR TITLE
[red-knot] Fix Leaking Narrowing Constraint in `ast::ExprIf`

### LINKED ISSUE #14588: [red-knot] Incorrect Narrowing Constraint Applied After ExprIf
```py
def bool_instance() -> bool:
    return True

x: Literal[42, "hello"] = 42 if bool_instance() else "hello"
reveal_type(x)  # revealed: Literal[42] | Literal["hello"]

_ = 1234 if isinstance(x, str) else 1234

# The `isinstance` test incorrectly narrows the type of `x`.
# As a result, `x` is revealed as Literal["hello"], but it should remain Literal[42, "hello"].
reveal_type(x)  # revealed: Literal["hello"]
```

This issue is not limited to `isinstance`. Other narrowing expressions or patterns also cause the same issue.

### PR DESCRIPTION
**Pull request by cake-monotone**
## Summary

Closes #14588

```py
x: Literal[42, "hello"] = 42 if bool_instance() else "hello"
reveal_type(x)  # revealed: Literal[42] | Literal["hello"]

_ = ... if isinstance(x, str) else ...

# The `isinstance` test incorrectly narrows the type of `x`.
# As a result, `x` is revealed as Literal["hello"], but it should remain Literal[42, "hello"].
reveal_type(x)  # revealed: Literal["hello"]
```

## Test Plan
mdtest included!

### CODE REVIEW

File: crates/red_knot_python_semantic/resources/mdtest/expression/if.md [hunk 1]
```diff
@@ -22,3 +22,22 @@ reveal_type(1 if None else 2)  # revealed: Literal[2]
 reveal_type(1 if "" else 2)  # revealed: Literal[2]
 reveal_type(1 if 0 else 2)  # revealed: Literal[2]
 ```
+
+## Leaked Narrowing Constraint
+
+(issue #14588)
+
+The test inside an if expression should not affect the scope outside of it.
```
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/resources/mdtest/expression/if.md:30]: micro-nit: `if` tests do not create new scopes in Python, unlike e.g. in Rust, so we have to use the vaguer term "block" here

```suggestion
The test inside an if expression should not affect code outside of the block.
```

**Review by AlexWaygood — APPROVED** [review]
LGTM (though David and Carl have done much more work in this area, so I'll wait for feedback from one of them before merging).

Great catch, thank you!

File: crates/red_knot_python_semantic/resources/mdtest/expression/if.md (hunk 1 above)
Comment (cake-monotone) [reply crates/red_knot_python_semantic/resources/mdtest/expression/if.md]: Thanks!!!! :)

**Review by carljm — APPROVED** [review]
Great catch, thank you!!

**Pull request merged**